h2o-extensions/target-encoder/src/main/java/ai/h2o/targetencoding/TEBroadcastJoin.java

```diff
@@ -0,0 +1,218 @@
+package ai.h2o.targetencoding;
+
+import water.MRTask;
+import water.MemoryManager;
+import water.fvec.CategoricalWrappedVec;
+import water.fvec.Chunk;
+import water.fvec.Frame;
+
+class TEBroadcastJoin {
+
+  /**
+   * @param leftFrame frame for which we want to keep the rows order.
+   * @param leftCatColumnsIdxs indices of the categorical columns from `leftFrame` for which we want to calculate encodings.
+   *        Only one categorical column is currently supported.
+   * @param leftFoldColumnIdx index of the fold column from `leftFrame` or `-1` if we don't use folds.
+   * @param rightFrame supposedly small frame that we will broadcast to all nodes and use as a lookup table for joining.
+   * @param rightCatColumnsIdxs indices of the categorical columns from `rightFrame`.
+   *        Only one categorical column is currently supported.
+   * @param rightFoldColumnIdx index of the fold column from `rightFrame` or `-1` if we don't use folds.
+   * @param maxFoldValue the highest fold value (e.g. 4 if 5 folds).
+   * @return the `leftFrame` with joined numerators and denominators.
+   */
+  static Frame join(Frame leftFrame, int[] leftCatColumnsIdxs, int leftFoldColumnIdx,
+                    Frame rightFrame, int[] rightCatColumnsIdxs, int rightFoldColumnIdx,
+                    int maxFoldValue) {
+    int rightNumeratorIdx = rightFrame.find(TargetEncoderHelper.NUMERATOR_COL);
+    int rightDenominatorIdx = rightNumeratorIdx + 1; // enforced by 
+    
+    // currently supporting only one categorical column
```

Comment:
Why do we have this limitation ? Can we get rid of it ? Not necessary in this PR - just a question.

Reply:
I guess it's just because TE doesn't need to join on more than one column for now.
It would make sense to join on multiple columns only if the encodings frame was created for 2 categorical columns at once, and I don't think anyone does this.
Another possibility is that the original author had something in mind for multiclass maybe, but I can't find any good alternative to the paper's proposal for such problems, so this is still a mystery...

Reply:
Will see in the future if I should just change the signature (avoid arrays if they're not used) or add support for join on multiple columns